### ISSUE TITLE
Publishing code coverage in builds takes a long time

### ISSUE THREAD

**Issue by mitchdenny** [Azure/azure-sdk-for-python#5293]
I think we should look into why publishing code coverage results to Azure DevOps is taking so long. Depending on the version of Python it seems to take between 7-9 minutes on each leg of the build. When we move to more scoped PR builds that time will be reduced, but even in the CI builds its a lot of time to spend publishing results.

I'm wondering whether for our purposes that just running code coverage on a single version of Python on a single machine would be sufficient. To what degree do we have special case handling per python runtime/OS that different code coverage results would have a meaningful impact?

**Comment by scbedd** [comment]
I believe that for the purposes of coverage, a single coverage run would probably be enough. However, we need to be aware that depending on the version + the platform, we _will_ be running slightly different code paths. Whether or not these slight differences have any impact on the greater coverage result should probably be discussed with the library team.

Why is publishing taking so long? Bunch of generated html files need to be uploaded to the artifacts. DevOps isn't the best with that.  We explicitly removed _our_ HTML genned reports (thanks @danieljurek !), but what remains is the upload of the summary.

**Comment by mitchdenny** [comment]
I was going to dig into that when this bubbled to the top of the stack. Many of the experiences in Azure DevOps which take in files rely on a feature called the File Container Service, and the libraries that tasks use to upload to this service aren't written with aggressive parallelization in mind. It can be more efficient to upload a single merged file than it can be to upload lots of little files (in the case of FCS).

I was going to bring it up with the test team and also have a look at the task implementation to see if an optimization can be made.

**Comment by mitchdenny** [comment]
Here is a related issue in the pipelines tasks repo:

https://github.com/microsoft/azure-pipelines-tasks/issues/4945

It's quite old but still seems relevant.

**Comment by mitchdenny** [comment]
The issue seems to be this code:

https://github.com/microsoft/azure-pipelines-agent/blob/1500e376855708134b1662d17ceddfba9519f0ba/src/Agent.Worker/CodeCoverage/CodeCoveragePublisher.cs

```csharp
            var publishCCTasks = files.Select(async file =>
            {
                var browsableProperty = (browsable) ? bool.TrueString : bool.FalseString;
                var artifactProperties = new Dictionary<string, string> {
                    { ArtifactUploadEventProperties.ContainerFolder, file.Item2},
                    { ArtifactUploadEventProperties.ArtifactName, file.Item2 },
                    { ArtifactAssociateEventProperties.ArtifactType, ArtifactResourceTypes.Container },
                    { ArtifactAssociateEventProperties.Browsable, browsableProperty },
                };

                FileContainerServer fileContainerHelper = new FileContainerServer(_connection, projectId, containerId, file.Item2);
                await fileContainerHelper.CopyToContainerAsync(context, file.Item1, cancellationToken);
                string fileContainerFullPath = StringUtil.Format($"#/{containerId}/{file.Item2}");

                Build.BuildServer buildHelper = new Build.BuildServer(_connection, projectId);
                var artifact = await buildHelper.AssociateArtifact(_buildId, file.Item2, ArtifactResourceTypes.Container, fileContainerFullPath, artifactProperties, cancellationToken);
                context.Output(StringUtil.Loc("PublishedCodeCoverageArtifact", file.Item1, file.Item2));
            });
```

Effectively it is serializing the upload of the files.

**Comment by mitchdenny** [comment]
Closing this down. Now that we are on unified pipelines, the time to upload code coverage results in the python pipeline isn't that significant. For storage it is about 22 seconds, it could possibly be made faster but the pay off for the effort of getting the Azure Pipelines agent code updated and chasing it through to production probably isn't worth it, and that assumes we've got someone to champion it for us on the other side.

I think for now it is "good enough".
